routing/payment_session.go

```diff
@@ -363,3 +366,63 @@ func (p *paymentSession) RequestRoute(maxAmt, feeLimit lnwire.MilliSatoshi,
 		return route, err
 	}
 }
+
+// UpdateAdditionalEdge looks up a channel edge policy using the ChannelID from
+// the msg. If the policy is found, it then tries to update it with the
+// supplied value. It returns a boolean to indicate whether the update is meant
+// for additional edges.
+func (p *paymentSession) UpdateAdditionalEdge(msg *lnwire.ChannelUpdate,
+	pubKey *btcec.PublicKey) bool {
+
+	// Query for the policy, if not found, then this update is not for
+	// additional edges.
+	policy := p.getAddtionalEdgePolicy(
+		pubKey, msg.ShortChannelID.ToUint64(),
+	)
+	if policy == nil {
+		return false
+	}
+
+	// Ignore outdated message and return true to indicate that this update
+	// is meant for private edge although we won't apply it.
+	lastUpdate := time.Unix(int64(msg.Timestamp), 0)
+	if !lastUpdate.Before(policy.LastUpdate) {
+		log.Warnf(
+			"Ignoring outdated update for private edges:%v",
+			policy.ChannelID,
+		)
+		return true
+	}
+
+	policy.TimeLockDelta = msg.TimeLockDelta
+	policy.FeeBaseMSat = lnwire.MilliSatoshi(msg.BaseFee)
+	policy.FeeProportionalMillionths = lnwire.MilliSatoshi(msg.FeeRate)
+	policy.LastUpdate = lastUpdate
+
+	log.Debugf("New private channel update applied: %v",
+		newLogClosure(func() string { return spew.Sdump(msg) }))
+
+	return true
+}
+
+func (p *paymentSession) getAddtionalEdgePolicy(pubKey *btcec.PublicKey,
+	channelID uint64) *channeldb.ChannelEdgePolicy {
+
+	target := route.NewVertex(pubKey)
+
+	for v, edges := range p.additionalEdges {
+		if v != target {
```

Comment:
We can do a map look up here (of the outer map) to avoid the outer loop iteration.